File: 1248-count-number-of-nice-subarrays/1248-count-number-of-nice-subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        ans = 0
        odds = 0
        
        l = 0
        for r in range(len(nums)):
            if nums[r] % 2:
                odds += 1
                
            if odds > k:
                l = first_odd_index + 1
                odds -= 1
            if odds == k:
                first_odd_index = l
                while nums[first_odd_index] % 2 == 0:
                    first_odd_index += 1
                
                ans += first_odd_index - l + 1
                
        return ans

File: 1248-count-number-of-nice-subarrays/1248-count-number-of-nice-subarrays.cpp
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        /*
            i need to know where it doesn't start the violation
            // if i am starting here
            // i need to know the position of odds that i have counted
            this means whenever i am poping a q i know my next index that i am not violating
        */
        for (auto &x: nums) {
            x &= 1;
        }
        int last = 0;
        int r = 0;
        int l = 0;
        int ans = 0;
        int cnt = 0;
        while (r < nums.size()) {
            if(nums[r]) {
                cnt++;
            }
            r++;
            while(cnt > k) {
                if(nums[l]) {
                    cnt--;
                }
                l++;
                last = l;
            }
            while(nums[last] == 0) {
                last++;
            }
            
            if(cnt == k) {
                ans += (last - l) + 1;
            }
        }
        return ans;
    }
};